Feature - Registration, User Profile view/edit and PIN reset

techaffinity opened the pull request:
The build that is associated with Registration feature includes user registration, user login, user profile view and edit/update and reset PIN. The update that covers validations and basic CSS alignment in the pages.

smn commented on ndohyep/app/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
[Tests seem to be missing?](http://ways-of-working.readthedocs.org/en/latest/process.html#our-development-process) In the mean time I'll get this repository setup with Travis & Coveralls.

smn commented:
Looping in @saeedmb8 and @bruskiza so they can follow along with the review process.

smn commented:
Travis failing because of errors coming from pep8 & pyflakes (using the flake8 tool).
Also lack of tests, could you add those please?

techaffinity commented:
Sure Simon. We will check these.

techaffinity commented:
We have fixed some of the errors in this update that has been run in our end and followed the standards. We will make sure for any other errors in the app and update the same as well.

smn commented:
Thanks, there are still flake8 errors making the travis build fail though. Check the [results](https://travis-ci.org/praekelt/molo-ndoh-yep/builds/72327816) for the specifics.

smn commented:
When this repository was created an extra (unnecessary) top level directory was added, I'm going to be removing that to make sure the setup.py and related project files are in the root of the directory.

smn commented:
Resolved merge conflict because this branch had diverged from `develop`. Merged develop back in and resolved conflicts.

smn commented:
Ok the build is now passing because I fixed all the remaining pep8 errors, it's also passing because:
```$ python manage.py test -v1
Creating test database for alias 'default'...
----------------------------------------------------------------------
Ran 0 tests in 0.000s
OK
Destroying test database for alias 'default'...
```
Looking forward to the tests. In the mean time I'll be reviewing the existing code without tests.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        widget=forms.TextInput(attrs=dict(required=True, max_length=30)),
+        label=_("Username"),
+        error_messages={'invalid': _("This value must contain only letters, \
+         numbers and underscores.")})
```
Your regex only allows words ('\w'):

``` python
>>> print re.match('^\w+$', 'a-_1')
None
>>> print re.match('^\w+$', 'abc')
<_sre.SRE_Match object at 0x1007c8920>
>>>
```

A test should catch this.

smn replied:
Also the `\` in the multiline string isn't necessary, since they're in brackets you can just use two strings on separate lines, they'll be combined into a single string:

``` python
>>> print ("foooo"
...        "bar"
...        "baz")
foooobarbaz
```

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+                max_length=4,
+                render_value=False,
+                type='password')),
+        label=_("PIN"))
```
Why call this field `password1`? Was this copied & pasted from a form that asked for the password twice for validation perhaps?

smn replied:
Why not make this a RegexField too?

``` python
>>> print re.match('^\d{4}$', 'asdf')
None
>>> print re.match('^\d{4}$', '12345')
None
>>> print re.match('^\d{4}$', '1234')
<_sre.SRE_Match object at 0x1007c88b8>
>>>
```

it would save you the `clean_password1()` function lower down.

techaffinity commented:
Sure Simon. We will check all these and update accordingly.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        label=_("Month"))
+    day = forms.IntegerField(
+        widget=forms.TextInput(attrs={'type': 'number'}),
+        label=_("Day"))
```
Would it make sense to make these ChoiceFields with sensible defaults? Would again save on validation and we can rely on the phone's own widget set which could arguably be a better UX.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+            User.objects.get(
+                username__iexact=self.cleaned_data['username']
+            )
+        except User.DoesNotExist:
```
Rather use 

``` python
if User.objects.exists(username__iexact=self.cleaned_data['username']):
    raise forms.ValidationError(_("Username already exists."))
return self.cleaned_data["username"]
```

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        length_of_day = len(str(day))
+        if length_of_day > 2 or day == 0:
+            raise forms.ValidationError(u'Please enter a valid day')
+        return day
```
clean_year, clean_month and clean_day can go once we use choices.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        return day
+
+    def clean(self):
+        return self.cleaned_data
```
I don't think this is necessary?

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        Otherwise we can allow the data to update.
+        """
+        cleaned_data = super(EditProfileForm, self).clean()
+        return cleaned_data
```
Is this a `TODO` item? If not then I think it can go away since `RegistrationForm()` already implements `clean()`

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        length_of_old_password = len(str(old_password))
+        if length_of_old_password > 4:
+            raise forms.ValidationError(u'Please enter a valid old password')
+        return old_password
```
Please use the `max_length` and `min_length` parameters available for the [CharField](https://docs.djangoproject.com/en/1.8/ref/forms/fields/#django.forms.CharField) type directly. This doesn't add any value over what Django already adds.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        if length_of_new_password > 4:
+            raise forms.ValidationError(u'New password should be \
+             maximum of 4 characters')
+        return new_password
```
Same as above wrt `max_length` and `min_length` on the CharField.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+        if length_of_confirm_password > 4:
+            raise forms.ValidationError(u'Confirm new password should be \
+             maximum of 4 characters')
+        return confirm_password
```
Same as above wrt `max_length` and `min_length` on the CharField.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+                return self.cleaned_data
+        except KeyError:
+            raise forms.ValidationError(_('Plase enter passwords \
+             in a valid format'))
```
When would this last one possibly be raised?

smn replied:
:point_up: please identify when this could happen or (of not or in doubt) rather remove it.

techaffinity replied:
The clean method has modified with removing of try, except block code.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,143 @@
+class ProfilePasswordChangeForm(forms.Form):
+    old_password = forms.CharField(widget=forms.PasswordInput())
+    new_password = forms.CharField(widget=forms.PasswordInput())
+    confirm_password = forms.CharField(widget=forms.PasswordInput())
```
All of these should have `required=True` and min & max length parameters set.

miltontony replied:
@smn `required` is `True` by default

smn commented on app/urls.py:
```diff
@@ -0,0 +1,30 @@
+from app.views import ProfilePasswordChangeView
+from django.conf.urls import patterns, url
+from django.contrib.auth.decorators import login_required
+# from django.conf import settings
```
Stray comment.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+                    password=form.cleaned_data['password1'],
+                )
+                profile = UserProfile(user=user, date_of_birth=dob)
+                profile.save()
```
Rather create the `UserProfile` using signals when a new `User` object is created. Allow for certain fields to be `null` and just update those `user.profile` fields here after the form validates.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+                )
+                profile = UserProfile(user=user, date_of_birth=dob)
+                profile.save()
+                return HttpResponseRedirect('/')
```
Please use `reverse()` to refer to named url paths.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+                profile.save()
+                return HttpResponseRedirect('/')
+            except ValueError:
+                messages.error(request, 'The day is out of range for month.')
```
Please put this `try`...`except` block only around the piece that can reasonable raise that error, which I think is the `datetime()` call on line 26.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+    variables = RequestContext(request, {
+        'form': form
+    })
+    return render_to_response(
```
Rather use an `else` for the `POST` check and create an empty `RegistrationForm` here. Results in less duplication of the `RequestContext` and `render_to_response()` calls.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+
+@login_required
+def home(request):
+    return HttpResponseRedirect('/')
```
`reverse()` and named url patterns for these please.

smn replied:
Also redirecting to `/` from a view called `home` seems a bit odd?

techaffinity replied:
We are working on this redirecting and will update on this specifically.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+from django.views.generic.edit import FormView
+from django.core.urlresolvers import reverse
+# from django.shortcuts import render
+from django.contrib import messages
```
Please remove stray comments.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+            context['alias'] = 'Anonymous'
+
+        context['date_of_birth'] = profile.date_of_birth
+        return context
```
Would a custom context processes that adds a `profile` key to all  template contexts be useful? See https://docs.djangoproject.com/en/1.8/ref/settings/#template-context-processors.

It would perhaps reduce the need for all this code here, you could solve that at the template level.

techaffinity replied:
We are checking this context-processors and will update on this.

techaffinity replied:
Simon, we tried to utilize this context-processors and it seems this was deprecated in django 1.8 and unable to customize the template context-processors. Please advice on the same.

miltontony replied:
Have a look at the [django 1.8 docs](https://docs.djangoproject.com/en/1.8/ref/templates/upgrading/#the-templates-settings) for setting context processors

techaffinity replied:
Sure Milton. We are checking.

smn replied:
I'm happy to help contribute this if needed.

techaffinity replied:
Simon, We have updated this one today. We had a problem with base.py after we modified the folder structure. We have added context processors in this latest update.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+            initial['alias'] = 'Anonymous'
+
+        initial['date_of_birth'] = profile.date_of_birth
+        return initial
```
See my comment above about a context processor that adds a `profile` to the default template context.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+
+    def get_form(self, form_class):
+        form = super(MyProfileEdit, self).get_form(form_class)
+        return form
```
Why the subclassing without any customisation?

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+
+    def get_form(self, form_class):
+        form = super(ProfilePasswordChangeView, self).get_form(form_class)
+        return form
```
Not needed I think.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+        else:
+            messages.error(
+                self.request,
+                'The Old password is incorrect.'
```
Please wrap in `_()` for i18n.

smn commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+        return render_to_response(
+            'app/change_password.html',
+            variables,
+        )
```
The second `return` from line 156 is unnecessary, it's already handled by the return on line 152 which is in the `else` block.

We can make this simpler by removing the else, moving everything beneath one indent to the left and removing lines 156 - 160.

smn commented on ndohyep/app/change_password.html:
```diff
@@ -0,0 +1,42 @@
+   <input type="submit" value='{% trans "Change Password" %}' class="btn btn-wide">
+</form>
+</div><!-- /form-wrap -->
+{% endblock %}
```
Missing newline.

smn commented on ndohyep/app/viewprofile.html:
```diff
@@ -0,0 +1,22 @@
+  </p>
+</div>
+
+{% endblock %}
```
Missing new line.

smn commented on ndohyep/registration/done.html:
```diff
@@ -0,0 +1,6 @@
+
+<p><b>{% trans "Thank you for joining" %}</b><br /></p>
+
+{% endblock %}
```
Missing new line.

smn commented on ndohyep/registration/login.html:
```diff
@@ -0,0 +1,18 @@
+    <div class="form-wrap">
+        <form method="post" action=".">{% csrf_token %}
+             {% if form.errors %}
+                <p class="errorlist">Your username and password does not match. Please try again.</p>
```
`{% trans "..." %}` please.

smn commented on ndohyep/registration/login.html:
```diff
@@ -0,0 +1,18 @@
+            {% endif %}
+            <table border="0">
+                <tr><th><label for="id_username">Username:</label></th><td>{{ form.username }}</td></tr>
+                <tr><th><label for="id_password">Password:</label></th><td>{{ form.password }}</td></tr>
```
Also `{% trans "..." %}` please.

smn commented on ndohyep/registration/login.html:
```diff
@@ -0,0 +1,18 @@
+            <input type="hidden" name="next" value="/home" />
+        </form>
+    </div><!-- /form-wrap -->
+    {% endblock %}
```
Missing new line.

smn commented on ndohyep/registration/login.html:
```diff
@@ -0,0 +1,18 @@
+                <tr><th><label for="id_password">Password:</label></th><td>{{ form.password }}</td></tr>
+            </table>
+            <input type="submit" value="Login" />
+            <input type="hidden" name="next" value="/home" />
```
Please use `{% url ... %}`

smn commented on ndohyep/registration/login.html:
```diff
@@ -0,0 +1,18 @@
+                <tr><th><label for="id_username">Username:</label></th><td>{{ form.username }}</td></tr>
+                <tr><th><label for="id_password">Password:</label></th><td>{{ form.password }}</td></tr>
+            </table>
+            <input type="submit" value="Login" />
```
`{% trans "..." %}`

smn commented on ndohyep/registration/register.html:
```diff
@@ -0,0 +1,24 @@
+        <input type="submit" class="registerSubmit" value="Register" />
+    </form>
+</div>
+{% endblock %}
```
missing new line.

smn commented on ndohyep/registration/register.html:
```diff
@@ -0,0 +1,24 @@
+        <table border="0">
+            {{ form.as_table }}
+        </table>
+        <input type="submit" class="registerSubmit" value="Register" />
```
`{% trans "..." %}`

smn commented on ndohyep/registration/register.html:
```diff
@@ -0,0 +1,24 @@
+{% endif %}
+
+<div class="form-wrap">
+    <form method="post" action=".">{% csrf_token %}
```
Why not post to a `{% url ... %}` ?

smn commented on ndohyep/registration/register.html:
```diff
@@ -0,0 +1,24 @@
+{% extends "base.html" %}
+{% load static core_tags %}
+{% block body_class %}template-{{ self.get_verbose_name|slugify }}{% endblock %}
```
Where is `self` defined? Is this a class based view thing?

smn replied:
@miltontony just told me that Wagtail makes it available in the template context.

smn commented on ndohyep/registration/success.html:
```diff
@@ -0,0 +1,9 @@
+{% extends "base.html" %}
+{% block title %}Registration Successful{% endblock %}
```
`{% trans "..." %}`

smn commented on ndohyep/registration/success.html:
```diff
@@ -0,0 +1,9 @@
+{% extends "base.html" %}
+{% block title %}Registration Successful{% endblock %}
+{% block head %}
+ Registration Completed Successfully
```
`{% trans "..." %}`

smn commented on ndohyep/registration/success.html:
```diff
@@ -0,0 +1,9 @@
+{% endblock %}
+{% block content %}
+    Thank you for registering.
+    <a href="/">Login</a> 
```
`{% trans "..." %}` and `{% url ... %}`

smn commented on ndohyep/registration/success.html:
```diff
@@ -0,0 +1,9 @@
+{% block content %}
+    Thank you for registering.
+    <a href="/">Login</a> 
+{% endblock %}
```
Missing new line.

smn commented on ndohyep/settings/dev.py:
```diff
@@ -7,7 +7,7 @@
+# try:
+#     from .local import *
+# except ImportError:
+#     pass
```
Are you not finding this useful or is this because of pyflakes complaining? I added an ignore entry to the `flake8` block in the `setup.cfg` file, it now no longer complains about `import *` statements.

smn commented on ndohyep/settings/production.py:
```diff
@@ -42,7 +42,7 @@
+# try:
+#     from .local import *
+# except ImportError:
+#     pass
```
See comment above about `import *` statements and flake8 checks.

smn commented on ndohyep/templates/base.html:
```diff
@@ -47,6 +47,20 @@
+                {% if user.profile.alias %}
+                    <strong>{{ user.profile.alias }}</strong> 
+                {% else %}
+                    <strong>Anonymous</strong> 
```
`{% trans "..." %}`

smn commented on ndohyep/templates/base.html:
```diff
@@ -47,6 +47,20 @@
+                    <strong>Anonymous</strong> 
+                {% endif %}
+            </div>
+            <li><a href="{% url 'view_my_profile' %}">My profile</a></li>
```
`{% trans "..." %}`

smn commented on ndohyep/templates/core/article_page.html:
```diff
@@ -1,4 +1,4 @@
-{% extends "base.html" %}
+{% extends "core/_article_page.html" %}
```
Can you explain the reasoning behind subclassing this extra template instead of `base.html` ?

techaffinity replied:
Simon, we have not modified this part of code we believe. Please advice.

smn replied:
Peculiar, ok, let's leave it as is then and I'll check with @miltontony to see if we made these changes or not.

smn commented:
ok, review done, mostly smallish things around missing translations, not consistently using the `{% url %}` and `reverse()` helpers to ensure decoupling from URLs and some suggestions on how to better use the existing Django form validation stuff instead of having to do a lot of the `clean_*()` form stuff manually.

techaffinity commented:
Thanks Simon. We have checked all these changes that has to be corrected from our end. We will modify and update the same. We will follow this kind of changes for other set of features we develop and make sure we dont have face any repetition in the upcoming modules development.

miltontony commented on app/views.py:
```diff
@@ -0,0 +1,160 @@
+        variables = RequestContext(request, {
+            'form': form
+        })
+        return render_to_response(
```
rather use [render](https://docs.djangoproject.com/en/1.8/topics/http/shortcuts/#render) here `render(request, 'registration/register.html', {'form': form})`. It avoids having to use the RequestContext

smn commented:
Excellent news, thanks team!

Please let me know when you're expecting to start adding the tests so I can plan another round of reviews.

We're anticipating this deliverable to be finished because we have work we want to be doing on another project using this.

techaffinity commented:
Simon, we have been working in polls feature and we had some good progress in setting up the models and Wagtail pages to support for the polls creation. We have noted all these changes and planned to fix these by Monday and could even update some test coverage by Monday on the same. We have noted some changes you have posted are have to be strongly followed in the upcoming modules and avoid repetitions.

smn commented:
Most of the things I've mentioned are either standard Django practices or things that we've lined out as expectations in our [ways of working](http://ways-of-working.readthedocs.org/) documentation. Please shout if there's any confusion about any of these issues.

smn commented:
Great, thanks @techaffinity.

techaffinity commented:
Definitely Simon. We will make sure for the Django practices and will push the code only after we fix the errors raised through flake8 and pep8 standards. We have some checklist now to follow up before pushing the code.

techaffinity commented:
Simon, we have updated the latest build in here with all these comments covered except for few that are related to context-processors. We are checking for those and will update on the same.

We have updated the test cases too in this build and checked for flake8 standards.

techaffinity commented:
Simon, when we tried pulling the latest code from this branch and while tried running, we could see the css, img and js folder was missing in the static folder in the root of molo-ndoh-yep/ndohyep. It throws error while running it. We have just placed the missing folders in the static and we could run the app then without any error. We have not pushed the static missing folders to the latest update we did yesterday. Please advice on the same.

techaffinity commented:
Simon, in our older version we had app folder inside the project where we have created polls app separately for the polls feature. Now we have changed the folder structure by bringing the app folder to the root. Please let me know whether we should have to keep polls app as separate or to write all the polls related models to the app folder in order to have app folder as common for all the modules. Please confirm this so that we could modify the structure in our end and we could push then for the first level update on the polls feature.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,96 @@
+
+
+class RegistrationForm(forms.Form):
+    this_year = datetime.date.today().year
```
Since this is a class level attribute I think it's only set at load time.
That means that if an app is started in `2015` and is never restarted in `2016` it will never show `2016`.
I think we should set this value in line in line 36 instead of setting a class attribute.

techaffinity replied:
Simon, we believe that we can get the current year from the datetime library once if we execute the view layer in the app like hitting the URL in the browser. We are not sure about your comment over restart. Please advice on the same.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,96 @@
+# from PIL import Image
+from django.contrib.auth.models import User
+from django.utils.translation import ugettext_lazy as _
+# from django.core.exceptions import ValidationError
```
Please remove all commented code or state very clearly with a note as to why it should remain here.

techaffinity replied:
Commented code has been removed.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,96 @@
+        return self.cleaned_data['username']
+
+    def clean_date_of_birth(self):
+        return self.cleaned_data['date_of_birth']
```
What is this needed for? It doesn't do anything I think?

smn commented on app/forms.py:
```diff
@@ -0,0 +1,96 @@
+        return self.cleaned_data['date_of_birth']
+
+    def clean(self):
+        return self.cleaned_data
```
Same for this, I don't think it does anything useful?

techaffinity replied:
The clean method code has been removed since it was unuseful.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,96 @@
+        del self.fields['date_of_birth']
+        self.fields.keyOrder = [
+            'alias',
+        ]
```
What use is this if this is the only field?

techaffinity replied:
This **init** method has been removed and used the django forms instead of RegistrationForm.

smn commented on app/forms.py:
```diff
@@ -0,0 +1,96 @@
+        widget=forms.PasswordInput(),
+        max_length=4,
+        min_length=4
+    )
```
I think these still need to be turned into RegexFields with a `PasswordInput()` widget. The `regex` should be the same as the one used for the registration form.

techaffinity replied:
The regex has been added to the fields.

smn commented on app/tests.py:
```diff
@@ -0,0 +1,50 @@
+        form_data = {
+            'old_password': '1234',
+            'new_password': 'jey3',
+            'confirm_password': 'jey3',
```
These are invalid passwords according to the spec, I think they're only allowed to be 4 digits.

smn commented on app/views.py:
```diff
@@ -0,0 +1,115 @@
+from django.views.decorators.csrf import csrf_protect
+from django.http import HttpResponseRedirect
+from app.models import UserProfile
+# from django import forms
```
Please remove commented code.

techaffinity replied:
Commented code has been removed.

smn commented on app/views.py:
```diff
@@ -0,0 +1,115 @@
+            )
+            profile = UserProfile(user=user, date_of_birth=dob)
+            profile.save()
+            return HttpResponseRedirect('/')
```
please use `reverse()`

smn commented:
@techaffinity please keep the modules separated and include them using `INSTALLED_APPS` in the settings file.

I see there are two `app` folders now, one in the root and one in the `ndohyep` folder. Is that perhaps a configuration issue gone a bit funny around how templates are located?

smn commented:
Once you've attended to the changes I'm :+1: on you merging. This PR is really too big to continue to review properly. Please try and make your PRs and your individual commits smaller to make reviews easier going forward.

smn commented:
:+1: please merge.

smn commented:
I'm going to go ahead and merge this in so I can review #5.
